tool/teleport/common/agentless.go

```diff
@@ -0,0 +1,724 @@
+		return trace.Wrap(err)
+	}
+
+	fmt.Printf("Writing Teleport keys to %s\n", clf.OpenSSHKeysPath)
+	if err := writeKeys(clf.OpenSSHKeysPath, newKeys); err != nil {
+		if a.defaultKeysDir == clf.OpenSSHKeysPath {
+			rmdirErr := os.RemoveAll(a.defaultKeysDir)
+			if rmdirErr != nil {
+				return trace.NewAggregate(err, rmdirErr)
+			}
+		}
+		return trace.Wrap(err)
+	}
+
+	fmt.Println("Updating OpenSSH config")
+	if err := updateSSHDConfig(
+		clf.OpenSSHConfigPath,
+		agentlessSSHDConfigPath,
+		filepath.Join(clf.OpenSSHKeysPath, teleportOpenSSHCA),
+		configuredHostKeys,
+		configuredHostCerts,
+	); err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+func (a *agentless) existingKeysPresent(clf config.CommandLineFlags) (bool, error) {
+	_, err := os.Stat(clf.OpenSSHKeysPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, trace.Wrap(err)
+	}
+	return true, nil
+}
+
+// openSSHInitialJoin
+func (a *agentless) openSSHInitialJoin(ctx context.Context, clf config.CommandLineFlags) error {
+	existing, err := a.existingKeysPresent(clf)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	privateKey, sshPublicKey, tlsPublicKey, err := GenerateKeys()
+	if err != nil {
+		return trace.Wrap(err, "unable to generate new keypairs")
+	}
+
+	var client auth.ClientI
+	var certs *proto.Certs
+
+	if existing {
+		exPrivateKey, _, _, exCerts, err := ReadKeys(clf)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		identity, err := auth.ReadIdentityFromKeyPair(exPrivateKey, exCerts)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		// todo(amk): if this or generate host certs fails, should we try to blowaway /etc/teleport/agentless and start from scratch in else section?
+		client, err = authenticatedUserClientFromIdentity(ctx, clf.InsecureMode, clf.FIPS, *a.proxyAddr, identity)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer client.Close()
+		certs, err = client.GenerateHostCerts(ctx, &proto.HostCertsRequest{
+			HostID:               a.uuid,
+			NodeName:             a.hostname,
+			Role:                 types.RoleNode,
+			AdditionalPrincipals: a.principals,
+			PublicTLSKey:         tlsPublicKey,
+			PublicSSHKey:         sshPublicKey,
+		})
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	} else {
+		certs, err = a.register(clf, sshPublicKey, tlsPublicKey)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		identity, err := auth.ReadIdentityFromKeyPair(privateKey, certs)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		client, err = authenticatedUserClientFromIdentity(ctx, clf.InsecureMode, clf.FIPS, *a.proxyAddr, identity)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		defer client.Close()
+	}
+
+	openSSHCA, err := a.getOpenSSHCA(ctx, client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	hostSSHCA, hostTLSCA, err := a.getHostCA(ctx, client)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```

Comment:
Why are you fetching these again? Don't you get the CAs as part of `certs` when you register or generate host certificates?